src/storage/store/var_list_column.cpp

```diff
@@ -39,39 +73,85 @@ void VarListColumn::scan(Transaction* transaction, node_group_idx_t nodeGroupIdx
+        for (auto i = 0u; i < numValues; i++) {
+            offset_t startOffset = listOffsetInfoInStorage.getListStartOffset(i);
+            offset_t appendLen = listOffsetInfoInStorage.getListLength(i);
+            KU_ASSERT(appendLen >= 0);
+            dataColumn->scan(transaction, nodeGroupIdx, startOffset, startOffset + appendLen,
+                dataVector, offsetToWriteListData);
+            offsetToWriteListData += appendLen;
+        }
+    }
 }
 
 void VarListColumn::scan(Transaction* transaction, node_group_idx_t nodeGroupIdx,
     kuzu::storage::ColumnChunk* columnChunk, offset_t startOffset, offset_t endOffset) {
     if (nodeGroupIdx >= metadataDA->getNumElements(transaction->getType())) {
         columnChunk->setNumValues(0);
     } else {
-        Column::scan(transaction, nodeGroupIdx, columnChunk, startOffset, endOffset);
-        // TODO: FIX-ME.
         auto varListColumnChunk = ku_dynamic_cast<ColumnChunk*, VarListColumnChunk*>(columnChunk);
-        auto startVarListOffset = varListColumnChunk->getListOffset(0);
-        auto endVarListOffset = varListColumnChunk->getListOffset(columnChunk->getNumValues());
-        auto numElements = endVarListOffset - startVarListOffset + 1;
-        varListColumnChunk->resizeDataColumnChunk(std::bit_ceil(numElements));
-        dataColumn->scan(transaction, nodeGroupIdx, varListColumnChunk->getDataColumnChunk(),
-            startVarListOffset, endVarListOffset);
+        Column::scan(transaction, nodeGroupIdx, columnChunk, startOffset, endOffset);
+        auto sizeColumnChunk = varListColumnChunk->getSizeColumnChunk();
+        sizeColumn->scan(transaction, nodeGroupIdx, sizeColumnChunk, startOffset, endOffset);
+        auto resizeNumValues = varListColumnChunk->getDataColumnChunk()->getNumValues();
+        bool isOffsetSortedAscending = true;
+        offset_t prevOffset = varListColumnChunk->getListStartOffset(0);
+        for (auto i = 0u; i < columnChunk->getNumValues(); i++) {
+            auto currentEndOffset = varListColumnChunk->getListEndOffset(i);
+            auto appendLen = varListColumnChunk->getListLen(i);
+            prevOffset += (uint64_t)appendLen;
+            if (currentEndOffset != prevOffset) {
+                isOffsetSortedAscending = false;
+            }
+            resizeNumValues += appendLen;
+        }
+        if (isOffsetSortedAscending) {
+            varListColumnChunk->resizeDataColumnChunk(std::bit_ceil(resizeNumValues));
+            offset_t startVarListOffset = varListColumnChunk->getListStartOffset(0);
+            offset_t endVarListOffset =
+                varListColumnChunk->getListStartOffset(columnChunk->getNumValues());
+            dataColumn->scan(transaction, nodeGroupIdx, varListColumnChunk->getDataColumnChunk(),
+                startVarListOffset, endVarListOffset);
+            varListColumnChunk->resetOffset();
+        } else {
+            varListColumnChunk->resizeDataColumnChunk(std::bit_ceil(resizeNumValues));
```

Comment:
This code path is not well covered by tests. Can you add tests that first perform a bunch of in place updates, then perform reads?